### PR TITLE
MQTT JSON respone for topic "/v"

### PR DESCRIPTION
**Pull request by Def3nder**
As a result of #615 I added a compile option to used JSON in the response to the "/v" topic instead of XML.
As the JSON "state" serialization does also include all segments, this response could grow to about 4k (and the segments are not included in the XML) so I changed the `serializeState` function to have an argument `bool includeSegments` to switch between "with/without segments".

I would suggest to add this to the MQTT settings later...

What do you think ?

### CODE REVIEW

**Comment by Def3nder** [comment]
Hi @Aircoookie ,

the implementation does do what is expected, however **the heap constantly decreases** 😢 
So I kind of mirrored the code from the WLED19_JSON.INO, but it does allocate memory everytime.

**So, please do not commit.**

**Comment by Aircoookie** [comment]
Hi @Def3nder !
It is a really good idea to support the JSON API via MQTT and something I should have added support for long ago.
Good thing you noticed the memory leak, I will try to fix it soon :) Maybe we can also find a way to include the segments, maybe in a different topic.

**Comment by Def3nder** [comment]
Hi @Aircoookie, I wonder why the memory leak is only there in this "add-on" and not in the JSON API.
I tested the JSON API with a flood of requests and the heap did stay at the samd level.

**Comment by Def3nder** [comment]
Hi @Aircoookie, I deleted to allocated JSON-object after using it - now the leak is away 😄 
Should be safe to commit now.

...something similar should be included in the WLED19_JSON.INO as well, shouldn't it ?

It would suggest in function `serveJson`:

change from:
```
[...]
  response->setLength();
  request->send(response);
}
```
to:
```
[...]
  response->setLength();
  request->send(response);
  delete response; response = NULL;
}
```
?

I wounder why this was necessary, because up to now I thought that objects would be deleted automatically if the scope has been left ... strange 😣

**Comment by Def3nder** [comment]
Hi @Aircoookie, I just thought about publishing the segments information:

if we push this to the MQTT broker each single UI change, this could lead to a delay.
I generated a dummy JSON with the max. amount of segments and with (looking into the future) 5 bytes per color. This is **2244 bytes** long (2030 of this are the segments):
[WLED_Status_incl_segments.json](https://github.com/Aircoookie/WLED/files/4117701/WLED_Status_incl_segments.json.txt)

What do you think ? Should I write some code to include the segment information maybe as "/seg" in the MQTT response ?

**Comment by Aircoookie** [comment]
Hi @Def3nder thank you for trying that out! Yes, I believe over 2k might be too much. I remember the MQTT homeassistant autodiscovery, it was 2.4kB and caused reboot loops and/or MQTT disconnects...

Another possible approach would be to have a topic for every single segment and only publish to the a segment's topic if that segment was updated. This might be more viable memory wise, but we would need to test if subscribing to so many topics at once could cause issues.

**Comment by Def3nder** [comment]
Hi @Aircoookie, or we do not publish segment changes and only offer the posibility to request the segment via MQTT. So the other system could send and /WLED/light-name/seg[0..9] and will get the content of one single segment as JSON in return.

What about this ?

File: wled00/wled19_json.ino [hunk 1]
```diff
@@ -191,8 +191,12 @@ void serializeSegment(JsonObject& root, WS2812FX::Segment& seg, byte id)
 	root["rev"] = seg.getOption(1);
 }
 
-
 void serializeState(JsonObject root)
+{
+  serializeState(root, true);
```
Comment (psi-4ward) [inline wled00/wled19_json.ino:196]: I had to move this call below it's definition to work

**Comment by Aircoookie** [comment]
Closing since outdated.
Still I hope to get around to adding MQTT JSON sending soon, one topic for info, one for state without segments, and one for each (updated) segment :)

**Pull request closed without merging**